[PT-D][Sharding] Move Partial Tensor to the _shard folder and add logic to remove padding

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76199

Since Partial Tensor is somehow isolated to sharded tensor. We now move it to the _shard folder.

Also, we added the logic to remove paddings when the size is not divisible by the world size. Modify the unit test to reflect this changes.

Finally, we need to consider the placement order for the reshading spec for partial tensor, related logic is added in this change. Futhermore, for sharded linear, we will need to order the placement by rank to get the expected local result.

Differential Revision: [D35827894](https://our.internmc.facebook.com/intern/diff/D35827894/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35827894/)!

Commit:
[PT-D][Sharding] Move Partial Tensor to the _shard folder and add logic to remove padding

Since Partial Tensor is somehow isolated to sharded tensor. We now move it to the _shard folder.

Also, we added the logic to remove paddings when the size is not divisible by the world size. Modify the unit test to reflect this changes.

Finally, we need to consider the placement order for the reshading spec for partial tensor, related logic is added in this change. Futhermore, for sharded linear, we will need to order the placement by rank to get the expected local result.

Differential Revision: [D35827894](https://our.internmc.facebook.com/intern/diff/D35827894/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35827894/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/partial_tensor.py:
```diff
@@ -127,25 +126,49 @@ def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> ShardedTensor:
-        local_shards = self.local_shard.chunk(
-            self.process_group.size(), dim=sharding_dim
-        )
+        current_rank = dist.get_rank(self.process_group)  # type: ignore[attr-defined]
+        rank_idx = None
+        rearrange_local_shards = False
+        indices = [0] * self.process_group.size()
+        for idx, placement in enumerate(resharding_spec.placements):  # type: ignore[attr-defined]
+            if placement.rank() == current_rank:  # type: ignore[index]
```
Is this needed because the existing code didn't work with rearrangements? If so, is there a unit test covering this case now?

fduwjj replied:
Previously we didn't consider that.. And yes, I added new unit test cases for it in test_partial_tensor.

pritamdamania87 commented on torch/distributed/_shard/partial_tensor.py:
```diff
@@ -127,25 +126,49 @@ def reshard(self, resharding_spec: shard_spec.ShardingSpec) -> ShardedTensor:
         )
 
         sharded_tensor_size = self.local_shard.size()
+        # Remove padding when the size is not divisible by the world size.
+        if chunk_mode_res != 0:
+            uneven_local_shards = self.local_shard.chunk(
+                self.process_group.size(), dim=sharding_dim
+            )
+            expected_size = uneven_local_shards[rank_idx].size()
+            if local_result.size() != expected_size:
+                local_result = local_result.narrow(
+                    sharding_dim,
+                    0,
+                    expected_size[sharding_dim],
+                )
```
Are there tests covering this case?

fduwjj replied:
Yes.

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/_ops/linear.py:
```diff
@@ -328,19 +325,16 @@ def _handle_row_wise_sharding_sharded_tensor(
-        indices[placement.rank()] = idx
-        if idx != placement.rank():
-            reaggrance_partial = True
+    if input.sharding_spec().dim not in (-1, len(input.size()) - 1):
+        raise NotImplementedError(
+            "The case when the input does not come from col-wise sharded "
+            "linear is not supported for row-wise sharded linear."
+        )
```
Does this end up breaking any existing supported use cases? Also, why do we need to remove the rearrangement logic we had before?

fduwjj replied:
This does not break any existing use case.

Because partial tensor does not contain any sharding info. So we use the natural order from rank 0 to N. And no rearrange is needed. I tested both sharded linear and megatron prototype, they both passed.

fduwjj commented on test/distributed/_shard/test_partial_tensor.py:
```diff
@@ -39,30 +39,33 @@ class TestPartialTensorReshard(ShardedTensorTestBase):
     def _run_partial_tensor_n_reshard(
         self, reshard_spec, input_size, world_size, reduce_op, dtype=torch.float
```
`resharding_spec` does come with different orders of placements and also there are cases when we need to do padding. So both cases are covered here. @pritamdamania87

Commit:
Update on "[PT-D][Sharding] Move Partial Tensor to the _shard folder and add logic to remove padding"


Since Partial Tensor is somehow isolated to sharded tensor. We now move it to the _shard folder.

Also, we added the logic to remove paddings when the size is not divisible by the world size. Modify the unit test to reflect this changes.

Finally, we need to consider the placement order for the reshading spec for partial tensor, related logic is added in this change. Futhermore, for sharded linear, we will need to order the placement by rank to get the expected local result.

Differential Revision: [D35827894](https://our.internmc.facebook.com/intern/diff/D35827894/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35827894/)!

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharding] Move Partial Tensor to the _shard folder and add logic to remove padding"


Since Partial Tensor is somehow isolated to sharded tensor. We now move it to the _shard folder.

Also, we added the logic to remove paddings when the size is not divisible by the world size. Modify the unit test to reflect this changes.

Finally, we need to consider the placement order for the reshading spec for partial tensor, related logic is added in this change. Futhermore, for sharded linear, we will need to order the placement by rank to get the expected local result.

Differential Revision: [D35827894](https://our.internmc.facebook.com/intern/diff/D35827894/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D35827894/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/_shard/sharded_tensor/ops/test_linear.py:
```diff
@@ -70,6 +70,7 @@ def _run_sharded_linear(
         inp = torch.rand(*input_size).cuda(self.rank)
         reshard_spec = copy.deepcopy(spec)
         reshard_spec.dim = 0
+        reshard_spec.placements.sort(key=lambda placement: placement.rank())
```
curious why we need to sort the placements here?

fduwjj replied:
Oh because we want to send back result of rank 0 to rank 0 because the placements are not always sorted by rank.
